Client/Frontend/Home/TopSitesPanel.swift

```diff
@@ -263,24 +228,57 @@ extension TopSitesPanel: UICollectionViewDelegate {
     }
 
     func collectionView(collectionView: UICollectionView, willDisplayCell cell: UICollectionViewCell, forItemAtIndexPath indexPath: NSIndexPath) {
-        if let thumbnailCell = cell as? ThumbnailCell {
-            thumbnailCell.delegate = self
-
-            if editingThumbnails && indexPath.item < dataSource.data.count && thumbnailCell.removeButton.hidden {
-                thumbnailCell.removeButton.hidden = false
-            }
-        }
+        let thumbnailCell = cell as? ThumbnailCell
+        thumbnailCell?.delegate = self
+        updateRemoveButtonStateForIndexPath(indexPath, forCell: thumbnailCell)
     }
 }
 
 extension TopSitesPanel: ThumbnailCellDelegate {
+
     func didRemoveThumbnail(thumbnailCell: ThumbnailCell) {
-        if let indexPath = collection?.indexPathForCell(thumbnailCell) {
-            if let site = dataSource[indexPath.item] {
-                self.deleteHistoryTileForSite(site, atIndexPath: indexPath)
+        guard let indexPath = collection?.indexPathForCell(thumbnailCell),
+            let site = dataSource[indexPath.item] else {
+            return
+        }
+
+        if !refillingTiles {
```

Comment:
`refillingTiles` gets set to true after deleting a tile and set back to false whenever `refillDeletedTiles()` finishes [1]. Refreshing the cache takes time and the user could try deleting a tile while we're refilling it. This ensures that won't happen. If we delete a tile while we're still refilling we end up having duplicates and the data source gets out of sync :(

https://github.com/mozilla/firefox-ios/pull/1283/files#diff-abfc8d88f8fbb974154bce59e22c81a6R164